### DISCUSSION TITLE
ffmpeg streaming usbcamera with mjpeg to  rtsp failed

### DISCUSSION THREAD

**Discussion by geekli** [bluenviron/mediamtx#1498] [Questions]
Ubuntu 22.04.2 LTS
rtsp-simple-server v0.21.4
Logitech StreamCam (type-c)
ffmpeg version 4.4.2-0ubuntu0.22.04.1  

`>$ lsusb`
> Bus 001 Device 004: ID 046d:0893 Logitech, Inc. StreamCam

`>$ sudo v4l2-ctl --list-devices`

>  Logitech StreamCam (usb-0000:00:14.0-4):
        /dev/video0
        /dev/video1
        /dev/media0

`>$ sudo ffmpeg  -hide_banner -f v4l2 -list_formats all -i /dev/video0`

> [video4linux2,v4l2 @ 0x55d133b24680] Raw       :     yuyv422 :           YUYV 4:2:2 : 640x480 176x144 320x240 424x240 640x360 848x480 960x540 1280x720 1600x896 1920x1080
[video4linux2,v4l2 @ 0x55d133b24680] Compressed:       mjpeg :          Motion-JPEG : 640x480 176x144 320x240 424x240 640x360 848x480 960x540 1280x720 1600x896 1920x1080
/dev/video0: Immediate exit requested

`>$ ffmpeg -hide_banner -encoders | grep -E "h264|mjpeg"`

>  V..... libx264              libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 (codec h264)
 V..... libx264rgb           libx264 H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 RGB (codec h264)
 V....D h264_nvenc           NVIDIA NVENC H.264 encoder (codec h264)
 V..... h264_omx             OpenMAX IL H.264 video encoder (codec h264)
 V..... h264_qsv             H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10 (Intel Quick Sync Video acceleration) (codec h264)
 V..... h264_v4l2m2m         V4L2 mem2mem H.264 encoder wrapper (codec h264)
 V....D h264_vaapi           H.264/AVC (VAAPI) (codec h264)
 V..... nvenc                NVIDIA NVENC H.264 encoder (codec h264)
 V..... nvenc_h264           NVIDIA NVENC H.264 encoder (codec h264)
 VFS... mjpeg                MJPEG (Motion JPEG)
 V..... mjpeg_qsv            MJPEG (Intel Quick Sync Video acceleration) (codec mjpeg)
 V....D mjpeg_vaapi          MJPEG (VAAPI) (codec mjpeg)

`>$ sudo ffmpeg -hide_banner -f v4l2 -input_format mjpeg -video_size 1280x720 -i /dev/video0 -vcodec copy  out1-2.mp4 `
> The output fale "out1-2.mp4" is normal

`>$ sudo ffmpeg -hide_banner -f v4l2 -input_format mjpeg -video_size 1280x720 -i /dev/video0 -vcodec copy  -f rtsp -rtsp_transport tcp   rtsp://192.168.2.98:8554/mystream`

> Input #0, video4linux2,v4l2, from '/dev/video0':
  Duration: N/A, start: 670344.083534, bitrate: N/A
  Stream #0:0: Video: mjpeg (Baseline), yuvj422p(pc, bt470bg/unknown/unknown), 1280x720, 30 fps, 30 tbr, 1000k tbn, 1000k tbc
Output #0, rtsp, to 'rtsp://192.168.2.98:8554/mystream':
  Metadata:
    encoder         : Lavf58.76.100
  Stream #0:0: Video: mjpeg (Baseline), yuvj422p(pc, bt470bg/unknown/unknown), 1280x720, q=2-31, 30 fps, 30 tbr, 90k tbn, 1000k tbc
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
Press [q] to stop, [?] for help
[rtp @ 0x56314b9ea3c0] RFC 2435 suggests two quantization tables, 3 provided
[rtp @ 0x56314b9ea3c0] RFC 2435 suggests two quantization tables, 3 provided
    Last message repeated 15 times
[rtp @ 0x56314b9ea3c0] RFC 2435 suggests two quantization tables, 3 provided
    Last message repeated 14 times
[rtp @ 0x56314b9ea3c0] RFC 2435 suggests two quantization tables, 3 provided
    Last message repeated 7 times
frame=   40 fps= 31 q=-1.0 Lsize=N/A time=00:00:01.40 bitrate=N/A speed=1.09x
video:5110kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: unknown
[video4linux2,v4l2 @ 0x56314b9de800] Some buffers are still owned by the caller on close.
ioctl(VIDIOC_QBUF): Bad file descriptor

`>$./rtsp-simple-server`

> 2023/02/21 13:33:19 INF rtsp-simple-server v0.21.2
2023/02/21 13:33:19 INF [RTSP] listener opened on :8554 (TCP), :8000 (UDP/RTP), :8001 (UDP/RTCP)
2023/02/21 13:33:19 INF [RTMP] listener opened on :1935
2023/02/21 13:33:19 INF [HLS] listener opened on :8888
2023/02/21 13:33:19 INF [WebRTC] listener opened on :8889 (HTTP)
2023/02/21 13:33:23 INF [RTSP] [conn 192.168.2.98:54868] opened
2023/02/21 13:33:23 INF [RTSP] [session 07278a0a] created by 192.168.2.98:54868
2023/02/21 13:33:23 INF [RTSP] [session 07278a0a] is publishing to path 'mystream', with TCP, 1 track (M-JPEG)
it seems all right,but after view http://192.168.2.98:8888/mystream/
INF [HLS] [muxer mystream] created (requested by 123.232.160.69)
2023/02/21 13:34:42 INF [HLS] [muxer mystream] destroyed (the stream doesn't contain any supported codec (which are currently H264, H265, MPEG4-Audio, Opus))
2023/02/21 13:34:45 INF [HLS] [muxer mystream] created (requested by xxx.xxx.xxx.xxx)
2023/02/21 13:34:45 INF [HLS] [muxer mystream] destroyed (the stream doesn't contain any supported codec (which are currently H264, H265, MPEG4-Audio, Opus))

**Comment by aler9** [comment]
Hello, as written in the logs, you are publishing a video track encoded with M-JPEG, and a M-JPEG track can't be read with HLS.
Your camera natively supports only M-JPEG, therefore you have to re-encode them stream with a supported codec, like H264:

```
sudo ffmpeg -hide_banner -f v4l2 -input_format mjpeg -video_size 1280x720 -i /dev/video0 -pix_fmt yuv420p -c:v libx264 -preset ultrafast -b:v 600k -max_muxing_queue_size 1024 -f rtsp -rtsp_transport tcp   rtsp://192.168.2.98:8554/mystream
```

[a comment by an automated account is not shown]
